gnomad_qc/v4/annotations/generate_freq.py

```diff
@@ -0,0 +1,1307 @@
+    """
+    ht = ht.annotate(
+        ab_adjusted_freq=hl.if_else(
+            ht.freq[0].AF > af_threshold,
+            hl.map(
+                lambda f, g: hl.struct(
+                    AC=hl.int32(f.AC + g),
+                    AN=f.AN,
+                    homozygote_count=f.homozygote_count + g,
+                    AF=hl.if_else(f.AN > 0, (f.AC + g) / f.AN, hl.missing(hl.tfloat64)),
+                ),
+                ht.freq,
+                ht.high_ab_hets_by_group_membership,
+            ),
+            ht.freq,
+        )
+    )
+
+    return ht
+
+
+def create_high_ab_age_hists_expr(ht: hl.Table, age_group_key="sample_age_bin"):
+    """
+    Create histograms of high ab counts using age bins to account for high AB hets becoming hom alts.
+
+    :param ht: Hail Table containing age hists, AB annotation.
+    :param age_group_key: Age group key to use for age histogram.
+    :return: Hail struct containing age histogram of high ab counts.
+    """
+    non_range_entries = hl.set(["n_larger", "n_smaller"])
+    age_bins_indices = hl.sorted(
+        hl.enumerate(ht["freq_meta"], index_first=False)
+        .filter(lambda x: x[0].contains(age_group_key))
+        .map(lambda x: (x[0][age_group_key], x[1]))
+    )
+
+    age_bins_indices_dict = hl.dict(age_bins_indices)
+    age_bin_indices_no_edges = age_bins_indices.filter(
+        lambda x: ~non_range_entries.contains(x[0])
+    )
+    return hl.struct(
+        bin_freq=hl.starmap(
+            lambda x, y: ht.high_ab_hets_by_group_membership[y],
+            age_bin_indices_no_edges,
+        ),
+        n_smaller=ht.high_ab_hets_by_group_membership[
+            age_bins_indices_dict["n_smaller"]
+        ],
+        n_larger=ht.high_ab_hets_by_group_membership[age_bins_indices_dict["n_larger"]],
+    )
+
+
+def get_sample_age_bin(
+    sample_age: hl.expr.Int32Expression,
+    lower_bound: int = 30,
+    upper_bound: int = 80,
+    number_of_bins: int = 10,
+) -> hl.expr.StringExpression:
+    """
+    Get the age bin for a sample.
+
+    :param sample_age: Sample age.
+    :return: Sample age bin.
+    """
+    bin_size = (upper_bound - lower_bound) / number_of_bins
+    lower_bin = hl.int(
+        hl.floor((sample_age - lower_bound) / bin_size) * bin_size + lower_bound
+    )
+    upper_bin = hl.int(lower_bin + bin_size)
+
+    bin_label = hl.if_else(
+        sample_age < lower_bound,
+        "n_smaller",
+        hl.if_else(
+            sample_age >= upper_bound,
+            "n_larger",
+            hl.str(lower_bin) + "-" + hl.str(upper_bin),
+        ),
+    )
+
+    return hl.or_missing(hl.is_defined(sample_age), bin_label)
+
+
+def compute_age_hist(mt: hl.MatrixTable) -> hl.MatrixTable:
+    """
+    Compute age histograms for each variant.
+
+    :param mt: Input MT with age annotation.
+    :return: MatrixTable with age histogram annotations.
+    """
+    mt = mt.annotate_rows(**age_hists_expr(mt.adj, mt.GT, mt.age))
```

Comment:
```suggestion
    mt = mt.annotate_rows(**age_hists_expr(mt.adj, mt.GT, mt.age))
```
In v2, we created an age hist per subset. I did not include the logic for this with the non-ukb subset so it needs to be added in. We can either split the age bins in the freq aggs to be  ukb-age-bin and non-ukb-age-bin  or try to determine these numbers from the current ukb and age bin freq counts. I need to think if the second approach accurately captures the age bins in the non-ukb subset.